Migrate Product Collection from WC to WP iAPI

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

Continue Product Collection migration from WC iAPI to WP iAPI.

Changes include:
- Replace `wc-navigation-id` with `wp-router-region`
- Adjust the frontend logic to it
- Remove loading bar animation as it's coming natively from iAPI router.

Testing covered:
- client-side navigation
    - one PC block on a page
    - multiple PC blocks on a page
- disabling client-side navigation when 3rd party block is added
- prefetching on hover
- JS events sending (PC rendering and viewing product)
- animations (loading bar, scroll to the first row of products)

> [!IMPORTANT]  
> It's OK CI fails on this PR with the following problems:
> - Lint error related to Cart -> Cart is being handled separately
> - E2E tests failing related to Cart -> Cart is being fixed spearately
> - E2E test failing related to Mini Cart: `Shopper → Notices -> Shopper can add item to cart, and will not see a notice in the mini cart` - It's being fixed in separate PR: https://github.com/woocommerce/woocommerce/pull/55295
>
> This PR aims another feature branch where we collect effort migrating iAPI and not `trunk`!

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Smoke test Product Collection in terms of iAPI usage:
- Client-side navigation
    - one PC block on a page
    - multiple PC blocks on a page
- disabling client-side navigation when 3rd party block is added
- prefetching on hover
- JS events sending (PC rendering and viewing product)
- animations (loading bar, scroll to the first row of products)

<!-- End testing instructions -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

kmanijak committed:
Fix iAPI navigation in Product Collection

kmanijak committed:
Extract dataset keys to a variavle

kmanijak committed:
Remove the loading bar animation and rely on iAPI one

kmanijak committed:
Fix lint

kmanijak committed:
Refactor the dataset access

kmanijak committed:
Merge branch 'dev/migrate-to-wp-interactivity-54259' into dev/migrate-iapi-product-collection

kmanijak committed:
Fix lint

kmanijak committed:
Partially fix E2E test

kmanijak committed:
Merge branch 'dev/migrate-to-wp-interactivity-54259' into dev/migrate-iapi-product-collection

tjcafferkey commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:
```diff
@@ -81,53 +93,32 @@ function scrollToFirstProductIfNotVisible( wcNavigationId?: string ) {
+				return;
+			}
+
+			const ctx = getContext< ProductCollectionStoreContext >();
```
Nitpick: We could move this into the `if()` condition so we don't execute unnecessarily

kmanijak replied:
https://github.com/woocommerce/woocommerce/pull/55330/commits/e3cd2611fe0157a48ca150821024bd8145b4778d

tjcafferkey commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:
```diff
@@ -81,53 +93,32 @@ function scrollToFirstProductIfNotVisible( wcNavigationId?: string ) {
+
+			const ctx = getContext< ProductCollectionStoreContext >();
+
+			const routerRegionId = getRouterRegionId( ref );
```
Nitpick: Same with this, we could move this into the if() condition so we don't execute unnecessarily

kmanijak replied:
https://github.com/woocommerce/woocommerce/pull/55330/commits/e3cd2611fe0157a48ca150821024bd8145b4778d

tjcafferkey reviewed (approved):
These changes look good, just left a few comments.

One UX issue I've spotted is pagination in general when using this block on a page:

1. Add Product Collection block and select a collection with around 5 or more products
2. Decrease "Products Per Page" option to 2, increase "Max Pages to Show" to more than 1
3. Save and go to frontend.
4. There is no pagination, I have to specifically add the pagination block (which I may not even know exists) to get this to work with the above options.

Are we already aware of this?

kmanijak commented:
> One UX issue I've spotted is pagination in general when using this block on a page:
>
> Add Product Collection block and select a collection with around 5 or more products
> Decrease "Products Per Page" option to 2, increase "Max Pages to Show" to more than 1
> Save and go to frontend.
> There is no pagination, I have to specifically add the pagination block (which I may not even know exists) to get this to work with the above options.
> Are we already aware of this?

@tjcafferkey, it depends on which collection you choose:
- if "create your own", the Pagination is included by default. The assumption is this would show a browsable catalog in some form
- if any other collection like "On Sale Products" or "New Arrivals" the default settings are: show 1 page only, no pagination, 5 products. The assumption is this is used for promo purposes or as a banner and is not browsable by default:

<img width="352" alt="image" src="https://github.com/user-attachments/assets/19d380ad-300f-4729-835e-e008017be920" />

Does it make sense?

kmanijak committed:
Move variables to if statements where they are used and fix types

tjcafferkey commented:
I suppose this is unrelated to this PR so feel free to merge but FWIW it makes sense, I can see the thinking there. My thinking is that it's easier to remove a block you don't want than it is to add a block you can't find (they may not know it exists/know its name)

kmanijak committed:
Remove code with accessibility messgaes as it's handled natively by iAPI

kmanijak commented:
@luisherranz pointed out iAPI router handles accessibility messages natively so I removed this from Product Collection because the announcements were duplicated https://github.com/woocommerce/woocommerce/pull/55330/commits/4dcfe34d39e785b9b3507011cd0243adcf909277.

Also @tjcafferkey, Luis said he'll make a review of this PR on Monday so I'm not merging it yet. Do you mind taking it over to address Luis review if needed and merging otherwise?

luisherranz commented on plugins/woocommerce/src/Blocks/BlockTypes/ProductCollection/Renderer.php:
```diff
@@ -249,7 +249,7 @@ private function enable_client_side_navigation( $block_content ) {
 		// Add `data-wp-navigation-id to the product collection block.
```
```suggestion
		// Add `data-wp-router-region` to the product collection block.
```

tjcafferkey replied:
Fixed in https://github.com/woocommerce/woocommerce/pull/55330/commits/dd24599bc375087db1e15e015d44c51b6caab5b9

luisherranz commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:
```diff
@@ -50,14 +59,14 @@ const forcePageReload = ( href: string ) => {
  */
-function scrollToFirstProductIfNotVisible( wcNavigationId?: string ) {
-	if ( ! wcNavigationId ) {
+function scrollToFirstProductIfNotVisible( wpRouterRegionId?: string ) {
```
We do this in the Query block by simply moving the focus to the first element in the list, which is important for accessibility. When moving the focus, browsers also scroll if the element is not on the viewport.

https://github.com/WordPress/gutenberg/blob/7fbba99059313dbf00e43a5e0c2cf62885852f79/packages/block-library/src/query/view.js#L41-L43

Even if you want to keep the manual scroll, the change of focus for accessibility should be added. The problem is that the two scrolls can conflict.

tjcafferkey replied:
Nice, this simplifies things quite a bit! https://github.com/woocommerce/woocommerce/pull/55330/commits/dd2193f06807ba0b4ac545dc32fd29e15ba8de73

luisherranz commented on plugins/woocommerce-blocks/assets/js/blocks/product-collection/frontend.tsx:
```diff
@@ -17,13 +17,22 @@ export type ProductCollectionStoreContext = {
+const getNavDisabled = ( ref: HTMLElement ) => {
+	const routerRegionElement = getRouterRegion( ref );
+	return routerRegionElement?.dataset?.wpNavigationDisabled;
+};
```
Instead of using this method, we are now using the property `clientNavigationDisabled` of the `core/router` config. 

`wp_interactivity_config( 'core/router', array( 'clientNavigationDisabled' => true ) );`

https://github.com/WordPress/gutenberg/blob/7fbba99059313dbf00e43a5e0c2cf62885852f79/packages/block-library/src/query/index.php#L123

In the client, there is nothing to do. The router already looks at that property and, if it exists, does not use client-side navigation.

tjcafferkey replied:
I've implemented this fix in this commit https://github.com/woocommerce/woocommerce/pull/55330/commits/340ba477b36938ceee3e4877190f908abb1ee057

I'd appreciate it if you could double check this one to ensure I've understood correctly.

luisherranz reviewed (changes requested):
Looks good to me, although there are still a couple of things that can be improved.

tjcafferkey committed:
Update comment from navigation-id to router-region

tjcafferkey committed:
Replace scroll calculations and functionality with a simple focus()

tjcafferkey committed:
Rely on core/router clientNavigationDisabled instead of custom attribute

luisherranz committed:
Refactor link validation and focus handling in product collection navigation

luisherranz commented:
Thanks for the changes, Tom. I'm taking a deeper look at how everything is handled in the Product Collection and I'd like to make some adjustments before this is merged. I'll push once I've finished.

luisherranz committed:
Avoid using random values for data-wp-key

luisherranz committed:
Consolidate  an unnecessary function and fix PHP lint issue

luisherranz committed:
Don't parse/create unnecessary context

luisherranz committed:
Minor lint issue

luisherranz committed:
Don't overwrite context, as it is already been declared

luisherranz committed:
Consolidate all the directives in a single function

With so many functions, is difficult to know which directives are
being added and where. Also, it's unnecessary (and wasteful) to
initialize more than one HTML Tag Processor to add these directives.

luisherranz reviewed (approved):
That's it. Feel free to review my changes and merge.

tjcafferkey committed:
Merge branch 'dev/migrate-to-wp-interactivity-54259' into dev/migrate-iapi-product-collection

tjcafferkey committed:
Merge branch 'dev/migrate-to-wp-interactivity-54259' into dev/migrate-iapi-product-collection
